x/concentrated-liquidity/types/keyspace.md

```diff
@@ -0,0 +1,107 @@
+# CL keyspace
+
+This document defines the key space format used in CL, and guarantees that are expected to hold.
+It was created to more easily reason about key space bugs that we were facing.
+
+If || is seen outside of a `backtick` quotation, it means raw appending.
+
+We use `string encoding` of an integer to mean the variable length, base10 encoding of an integer into a string.
+
+Also all of this really shows we vastly overpay for complexity in the state machine, and that collections work in the SDK can really simplify this code.
+
+We break this up into parts, the multi-component keys, and the single component keys that don't really have concern.
+
+TODO: How do we train AI tools to do more of this for us, instead if it being manual
+- Maybe not worth doing since collections / ORM should delete this code for us.
+
+## multi-component keys
+
+## 0x01 - Pool Tick storage
+
+If a key exists in state, that begins with `0x01`, it is expected that it is of the form:
+`0x01` || `8 byte big endian encoding of pool ID` || `9 byte signed tick encoding`
+
+We are expected to be able to iterate over all ticks in a pool, from most negative to most positive.
+
+## 0x02 - Indexed position ID storage
+
+If a key exists in state, that begins with `0x02`, it is expected that it is of the form:
+`0x02/` || `Hexadecimal encoding of an address` || `/` || `string encoding of pool ID` || `/` || `string encoding of position ID`
+
+- We are expected to be able to safely iterate over all positions for an address
+    - Iterate over `0x02/` || `Hexadecimal encoding of an address` || `/` 
+- We are expected to be able to safely iterate over all positions for an address, pool_ID pairs.
+    - Iterate over `0x02/` || `Hexadecimal encoding of an address` || `/` || `string encoding of pool ID` || `/` 
+
+Since we encode the address in hexadecimal, and hexadecimal doesn't contain `/`, there is no malleability on the encoding.
+
+However, we are likely better off encoding this address with bech32, which also doesn't contain `/`, but has the benefit of being:
+- A more user friendly representation
+- Marginally more space inefficient. ("43" bytes vs "40" bytes)
```

Comment:
```suggestion
- Marginally more space efficient. ("43" bytes vs "40" bytes)
```

Reply:
no its more inefficient, the hex is 40 bytes, the bech32 is 43

Reply:
Ah, thought it was a typo it being "it has the benefit of having marginally more space inefficient"

Reply:
oh lol